overlay/usr/lib/inithooks/bin/nextcloud.py

```diff
@@ -27,59 +38,230 @@ def usage(s=None):
     sys.exit(1)
 
 
+def set_password(password: str) -> tuple[int, str, list[str]]:
+    """Set Nextcloud password
+    returns exit_code, message & list of failed services
+        exit_code is one of:
+            - SUCCESS
+            - ERR_PASSWORD
+            - ERR_EXCEPTION
+            - ERR_UNKNOWN
+    """
+    local_env = os.environ.copy()
+    local_env["OC_PASS"] = password
+    p = subprocess.run(
+        args=[
+            "/usr/local/bin/turnkey-occ",
+            "user:resetpassword",
+            "--password-from-env",
+            "admin",
+        ],
+        cwd="/var/www/nextcloud",
+        env=local_env,
+        text=True,
+        capture_output=True,
+    )
+    # Success :)
+    if p.returncode == SUCCESS:
+        log.write("admin password set successfully")
+        return (SUCCESS, "", [])
+
+    # password setting failure/error handling
+    else:
+        # occ seems to only output to stdout - even if stacktrace
+        # but just in case that changes
+        occ_output = p.stdout + p.stderr
+        services = []
+        # log raw occ output - including full stacktrace if relevant
+        log.write(occ_output, "err")
+
+        # password complexity failure
+        if "password" in occ_output.lower():
+            error_code = ERR_PASSWORD
+            msg = occ_output.rstrip()
+
+        # nextcloud exception
+        elif "exception" in occ_output.lower():
+            error_code = ERR_EXCEPTION
+            msg = "Nextcloud exception:"
+
+            if "redis" in occ_output.lower():
+                services.append("Redis")
+            if "database" in occ_output.lower():
+                services.append("MariaDB")
+            if services:
+                msg = f"{msg} Can't connect to service(s)"
+            else:
+                msg = f"{msg} Unknown exception"
+
+        # any other unknown/unexpected error
+        else:
+            error_code = ERR_UNKNOWN
+            msg = f"Unexpected Nextcloud error"
+
+        return (error_code, msg, services)
+
+
+def set_random_pass() -> tuple[int, str]:
+    """Returns exit_code & msg"""
+    chars = string.ascii_letters + string.digits
+    while True:
+        random_pass = ''.join(secrets.choice(chars) for _ in range(32))
+        if (
+            any(c.islower() for c in random_pass)
+            and any(c.isupper() for c in random_pass)
+            and sum(c.isdigit() for c in random_pass)
```

Comment:
This seems like an odd check to me:
`sum(c.isdigit() for c in random_pass)` will first, create an iterator of booleans for each character, convert them to integers (False => 0, True => 1), sum them together, then convert the integer to a boolean for the if check.

It's functionally equivalent to `any(c.isdigit() for c in random_pass)`, but less idiomatic imo.